storage/shard.go

```diff
@@ -199,6 +185,79 @@ func (s *dbShard) QueryRaw(
 	return res, nil
 }
 
+func (s *dbShard) QueryGrouped(
+	ctx context.Context,
+	q query.ParsedGroupedQuery,
+) (query.GroupedResults, error) {
+	s.RLock()
+	if s.closed {
+		s.RUnlock()
+		return query.GroupedResults{}, errShardAlreadyClosed
+	}
+
+	active, sealed, cleanup := s.getEligibleSegmentsWithLock(q.StartNanosInclusive)
+	s.RUnlock()
+
+	defer cleanup()
+
+	// Querying active segment and adds to result set.
+	activeRes, err := active.QueryGrouped(ctx, q)
+	if err == errMutableSegmentAlreadySealed {
+		// The active segment has become sealed before a read can be performed
+		// against it. As a result we should retry the read.
```

Comment:
Should we wait before retrying here?

Reply:
Also, why don't we just check the accessor count and delay sealing of the active segment until the accessor count goes to 0? Or are we afraid of query contention from delaying segment sealing?

Reply:
Not sure why we should wait before retrying? The error means the active segment has been changed so should be re-queried.

To prevent a race condition between checking accessor count and querying, you'll need a write lock to make sure the accessor count doesn't change while the segment is queried, no?

Reply:
Mmm that's true. Nvm then haha.